### DISCUSSION TITLE
ability to install nodejs versions without npm?

### DISCUSSION THREAD

**Discussion by jasisk** [jdx/mise#644] [Ideas]
I'm not entirely sure how the api would look to support this but I'm curious if it is possible to optionally install nodejs versions without npm.

This would enable two things, from what I can tell:
1. allow managing npm versions with `corepack`: right now, the corepack shim fails to install when attempting to enable for npm (`corepack enable npm`) from a node/corepack version managed by rtx.
2. allow managing npm from a different global npm store: setting a `prefix` in an `.npmrc` file allows you to move your global npm store to something like, `~/.npm-global`, allowing globally install modules to persist across versions (including npm itself).

**Comment by jdx** [comment]
npm comes included in the official tarballs, so it's not really possible to not include it, however I think the idea of specifying alternate npm/package manager locations/versions does make sense. I think that would make more sense as an override rather than simply not including npm though

**Reply by jasisk** [reply]
Makes sense to me.

If I might suggest one thing to consider that would address point (1): one difference between the shipped tarball and the `rtx` installation is that the `npm` script is a symlink in the tarball but the real file in `rtx`. This prevents `corepack` from being able to replace the `npm` symlink. Could a middle-ground be to leave the symlink in place, similar to that for `npx`?

**Reply by jdx** [reply]
That's done to shim npm so that it automatically creates shims when running `npm i -g`. I think we'll need a different solution to that problem before we can make it compatible with corepack. I suspect more people rely on the auto-shimming than want to use corepack.

**Reply by jdx** [reply]
perhaps what I could do is modify the npm cli directly rather than putting a shim in front of it. Or maybe I can make it a symlink to a shim instead of a shim directly (if that works with corepack, dunno)

**Reply by jasisk** [reply]
Ack. Didn't look at the contents of the `npm` file before commenting. 🤦 
You're absolutely right—the `corepack` shim is in conflict with the `rtx` shim. So it could be an either / or deal (either you don't `corepack enable npm` and you get `rtx` auto-shimming, or `enable npm` and lose `rtx` auto-shimming) but I can totally see if this is too confusing or inconsistent to want to implement.

**Comment by JamBalaya56562** [comment]
Both halves of this have answers now, and the corepack one is a supported setting.

**`node.npm_shim = false` turns the wrapper off, and mise then runs `corepack enable npm` for you** — exactly the either/or @jasisk landed on in this thread. Measured on **v2026.8.3**, two clean installs of node 24:

```console
# [settings.node] npm_shim = true (the default), corepack = false
$ mise install
mise node@24.19.0  ✓ installed
$ ls -l .../installs/node/24.19.0/bin/npm
-rwxr-xr-x 1 jam jam 2907 Aug  9 14:57 .../bin/npm
$ head -1 .../installs/node/24.19.0/bin/npm
#!/usr/bin/env bash

# [settings.node] npm_shim = false, corepack = true
$ mise install
mise node@24.19.0    [3/3] enable corepack npm shim
mise node@24.19.0  ✓ installed
$ ls -l .../installs/node/24.19.0/bin/npm
lrwxrwxrwx 1 jam jam 40 Aug  9 14:57 .../bin/npm -> ../lib/node_modules/corepack/dist/npm.js
$ ls .../installs/node/24.19.0/bin
corepack  node  npm  npx  pnpm  pnpx  yarn  yarnpkg
```

```toml
[settings.node]
npm_shim = false
corepack = true
```

@jasisk's diagnosis was right, and the PR that fixed it restates it almost verbatim — the wrapper is a real file where the upstream tarball ships a symlink, and corepack refuses to clobber a non-symlink `bin/npm`. That is **#10082**, first released in **v2026.5.16** (bisected: `npm_shim` is absent from `settings.toml` at v2026.5.15 and present at v2026.5.16). Per its description, this is what makes `packageManager: "npm@X.Y.Z"` in `package.json` take effect for npm rather than being silently ignored while yarn/pnpm work.

The trade-off is the one @jdx named here: with the wrapper off you lose the automatic `mise reshim` after `npm install -g`. Run `mise reshim` yourself, or use the `npm:` backend for CLI tools you want shimmed.

On point (2), the `prefix`-in-`.npmrc` case — with `npm_shim = false` mise no longer sits in front of npm, so a global prefix should behave as it does outside mise. I did not measure that half.
